lib/ClangImporter/ClangImporter.cpp

```diff
@@ -2732,36 +2732,41 @@ void ClangImporter::lookupTypeDecl(
   clang::DeclarationName clangName(
       &Impl.Instance->getASTContext().Idents.get(rawName));
 
-  clang::Sema::LookupNameKind lookupKind;
+  SmallVector<clang::Sema::LookupNameKind, 1> lookupKinds;
   switch (kind) {
   case ClangTypeKind::Typedef:
-    lookupKind = clang::Sema::LookupOrdinaryName;
+    lookupKinds.push_back(clang::Sema::LookupOrdinaryName);
     break;
   case ClangTypeKind::Tag:
-    lookupKind = clang::Sema::LookupTagName;
+    lookupKinds.push_back(clang::Sema::LookupTagName);
+    lookupKinds.push_back(clang::Sema::LookupNamespaceName);
     break;
   case ClangTypeKind::ObjCProtocol:
-    lookupKind = clang::Sema::LookupObjCProtocolName;
+    lookupKinds.push_back(clang::Sema::LookupObjCProtocolName);
     break;
   }
 
   // Perform name lookup into the global scope.
   auto &sema = Impl.Instance->getSema();
-  clang::LookupResult lookupResult(sema, clangName, clang::SourceLocation(),
-                                   lookupKind);
   bool foundViaClang = false;
-  if (!Impl.DisableSourceImport &&
-      sema.LookupName(lookupResult, /*Scope=*/nullptr)) {
-    for (auto clangDecl : lookupResult) {
-      if (!isa<clang::TypeDecl>(clangDecl) &&
-          !isa<clang::ObjCContainerDecl>(clangDecl) &&
-          !isa<clang::ObjCCompatibleAliasDecl>(clangDecl)) {
-        continue;
-      }
-      auto *imported = Impl.importDecl(clangDecl, Impl.CurrentVersion);
-      if (auto *importedType = dyn_cast_or_null<TypeDecl>(imported)) {
-        foundViaClang = true;
-        receiver(importedType);
+
+  for (auto lookupKind : lookupKinds) {
+    clang::LookupResult lookupResult(sema, clangName, clang::SourceLocation(),
+                                     lookupKind);
+    if (!Impl.DisableSourceImport &&
+        sema.LookupName(lookupResult, /*Scope=*/ sema.TUScope)) {
```

Comment:
This makes more sense. I wonder what it defaulted to with `Scope=nullptr` before.

Reply:
Before this change the C++ lookup logic skipped most of the actual lookup and returned here: https://github.com/apple/llvm-project/blob/7a112cc728541a3dc273abdc73c68f031d8d74bd/clang/lib/Sema/SemaLookup.cpp#L1415